spec.bs

```diff
@@ -447,6 +448,83 @@ To <dfn>build an interest group permissions url</dfn> given a [=origin=] |ownerO
+[=interest groups=]. It's required that [=interest group set=] has no more than:
+  * 2000 [=interest group/owners=];
+  * 1000 [=regular interest groups=] per [=interest group/owner=];
+  * 20000 [=negative interest groups=] per [=interest group/owner=];
+  * <code>10\*1024\*1024</code> total [=interest group/estimated size=] of [=interest groups=] per
+    [=interest group/owner=].
+
+<div algorithm='maintenance'>
+To <dfn>perform storage maintenance</dfn>:
+
+1. Let |ownersAndExpiry| be a new [=ordered map=] whose [=map/keys=] are [=origins=]
+  [=interest group/owner=] and [=map/values=] are [=moments=] [=interest group/expiry=].
+1. [=list/For each=] |ig| of the user agent's [=interest group set=]:
+  1. Let |owner| be |ig|'s [=interest group/owner=].
+  1. If |ownersAndExpiry|[|owner|] [=map/exists=], then [=map/set=] |ownersAndExpiry|[|owner|] to
+    |ig|'s [=interest group/expiry=] if it comes after |ownersAndExpiry|[|owner|].
+  1. Otherwise, [=map/set=] |ownersAndExpiry|[|owner|] to |ig|'s [=interest group/expiry=].
+1. If [=map/size=] of |ownersAndExpiry| |i| &gt; 2000, then set |ownersAndExpiry| to
+  |ownersAndExpiry| [=map/sorted in descending order=] with |a| being less than |b| if |a|'s
+  [=map/value=] comes before |b|'s [=map/value=].
+
+  Note: Sorting owners based on their expiry in descending order so that if interest groups of some
+    owners need to be removed, the interest groups of owners expiring soonest are the ones to be
+    removed first.
+
+1. Let |owners| be the [=map/get the keys|keys=] of |ownersAndExpiry|.
+1. [=list/For each=] |i| in [=the range=] from 0 to the [=set/size=] of |owners|, exclusive:
+  1. If |i| &ge; 2000, then [=list/remove=] [=interest groups=] from the user agent's
+    [=interest group set=] whose [=interest group/owner=] is |owners|[|i|], and [=iteration/continue=].
+  1. Let |regularIgs| be a [=list=] of [=regular interest groups=] in the user agent's
+    [=interest group set=] whose [=interest group/owner=] is |owners|[|i|].
+  1. If [=list/size=] of |regularIgs| &gt; 1000, then [=clear excess interest groups=] with
+    |regularIgs|, |owners|[|i|], and 1000.
+  1. Let |negativeIgs| be a [=list=] of [=negative interest groups=] in the user agent's
+    [=interest group set=] whose [=interest group/owner=] is |owners|[|i|].
+  1. If [=list/size=] of |regularIgs| &gt; 20000, then [=Clear excess interest groups=] with
+    |negativeIgs|, |owners|[|i|], and 20000.
+1. TODO: remove expired IGs.
+1. [=list/For each=] |owner| of |owners|:
+  1. Let |igs| be a [=list=] of [=interest groups=] in the user agent's [=interest group set=] whose
+    [=interest group/owner=] is |owner|, [=list/sorted in descending order=] with |a| being less
+    than |b| if |a|[=interest group/expiry=] comes before |b|[=interest group/expiry=].
+  1. Let |cumulativeSize| be 0.
+  1. [=list/For each=] |ig| of |igs|:
+    1. If the sum of |cumulativeSize| and |ig|'s [=interest group/estimated size=]

```

Comment:
Our code does not use the [=estimated size=], but [a different way for size calculation](https://crsrc.org/c/content/browser/interest_group/interest_group_storage.cc;drc=f5bdc89c7395ed24f1b8d196a3bdd6232d5bf771;l=3273) because it requires a lot of extra work in the code to use [=estimated size=] as the code comment explains. Not sure if we need to add a separate size calculation to sync with the code in the spec, or simply using [=estimated size=] is fine.

Reply:
See this [SQL code](https://crsrc.org/c/content/browser/interest_group/interest_group_storage.cc;drc=f5bdc89c7395ed24f1b8d196a3bdd6232d5bf771;l=3309) I linked above. It uses the `LENGTH` of some fields of the IG stored in DB, and then add a fixed 40 for remaining fields. For example, it does not have the size of [execution mode](https://crsrc.org/c/third_party/blink/common/interest_group/interest_group.cc;drc=977dc02c431b4979e34c7792bc3d646f649dacb4;l=279), a boolean field [enable_bidding_signals_prioritization](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/public/common/interest_group/interest_group.h;drc=977dc02c431b4979e34c7792bc3d646f649dacb4;bpv=1;bpt=1;l=109?gsn=enable_bidding_signals_prioritization&gs=KYTHE%3A%2F%2Fkythe%3A%2F%2Fchromium.googlesource.com%2Fcodesearch%2Fchromium%2Fsrc%2F%2Fmain%3Flang%3Dc%252B%252B%3Fpath%3Dthird_party%2Fblink%2Fpublic%2Fcommon%2Finterest_group%2Finterest_group.h%23Pcosih4sQ8BueDrOoNGTHmxSH5ctfCu0AV-M0kgu3sc), etc, which are counted in the fixed 40.

Reply:
Why do we do that? You mentioned that the code comment explains that it's a bunch of extra work, but I'm not sure what that extra work is, sorry if I'm missing it. I have three questions:

1. Does the spec currently do all of that extra work, instead of just adding an opaque "40" to cover all of the fields?
2. What are the web-observable consequences of another implementation coming along and using a different size calculation mechanism? Will this be web-observable in any way? Can it be tested with WPT?
3. What are the consequences of changing the size calculation in the future. I'm trying to estimate the possible compatibility consequences of this, if any. Would it be really hard for Chromium to change "40" into "50", or would that be a subtle change that has big consequences, and must only be considered carefully with customer feedback?